### PR TITLE
Don't allow start to equal end

### PR DESCRIPTION
**Pull request by artificial-aidan**

### CODE REVIEW

**Comment by vcschapp** [comment]
Thanks for this contribution, Aidan!

Before I look at the code, could you give a quick summary of what problem this is solving, and why this is the right way to solve it?

**Comment by artificial-aidan** [comment]
Fair warning, this was my first time using this, and I still don't have a complete understanding.

I found that with certain time ranges I ran into 2 problems. 

1. Start would sometimes equal end on the last chunk. The test I added catches this. Cloudwatch log insights fails if start equals end. This could maybe be fixed with logic around whether or not to advance to another chunk. But I tried this fix first. Could use input on this.
2. Sometimes when doing like a 2-3 hour query range with small splitUntil settings, I would end with a last chunk where start was before end. I wasn't able to reproduce this with a unit test, but maybe with more knowledge about how it works, you might have more insight.

Not sure this is the right way to solve it, but it's one way 😄

**Comment by vcschapp** [comment]
If I'm understanding correctly, it sounds like in some cases the chunking logic (or chunk splitting logic?) may be generating a "null chunk" in some cases or possibly an "inverted chunk"... Just spot-checking the code, it seems like a plausible bug, although I can't see how it materializes (yet).

- Could you share an input `QuerySpec` and the final error message for a case in which you're able to trigger this problem?
- I would feel more comfortable if we had a simple unit test that triggers the issue and acts as a validation of the fix and future regression test.
   - Do you think you could attempt this?
   - I'm thinking the place to put it would be `TestQueryManager_Query` in `mgr_test.go` probably as a new test along the lines of `t.Run("Query Does Not Cause Null or Inverted Chunk Time Ranges", ...)`
   - The trick with these unit tests is figuring out the minimum inputs and minimum mock service interactions needed to put you in the bad state.

**Comment by artificial-aidan** [comment]
So for issue #1, the unit test I added covers it. If you revert my code you will see start equals end.

For issue 2 I can take another try, the problem only manifested itself on like a 3 hour query window with a random start time, so it took 15 minutes to reproduce

**Comment by vcschapp** [comment]
Agreed for issue #1, that the unit test covers the proximate issue, *i.e.* prevent chunk end being before start.

What I'm hoping for is a test that covers the root cause, *i.e.* what set of inputs and/or service interactions would result in a code path being followed that that would cause chunk end to be before start. That's how we can see that there really is a bug.

(I can definitely help advise on this... I really want to get this fix in, but would like the behavior to be reproducible first.)

File: stream.go [hunk 1]
```diff
@@ -135,5 +135,8 @@ func (s *stream) nextChunkRange() (start, end time.Time) {
 	if start.Before(s.Start) {
 		start = s.Start
 	}
+	if start.Equal(s.End) || end.Before(start) {
```
Comment (vcschapp) [inline stream.go:138]: nit, but this line simplifies to `if !end.After(start)` ...

File: stream_test.go [hunk 2]
```diff
@@ -397,6 +397,21 @@ func TestStream_NextChunkRange(t *testing.T) {
...
 		},
+		{
+			name: "Start/End cannot be the same",
+			s: stream{
+				QuerySpec: QuerySpec{
+					Start: defaultEnd.Add(-3 * time.Hour),
+					End:   defaultEnd,
+					Chunk: time.Minute * 1,
+				},
+				next: 180,
```
Comment (vcschapp) [inline stream_test.go:408]: I think if I'm understanding correctly, this is trying to recreate a state that should not be possible if the rest of the Incite engine is working correctly.

Basically a 3-hour time range chunked in minutes has 180 chunks numbered `0` through `179`. Chunk number `180` is "one past the end" and so you would expect it to have odd characteristics.

The questions in my mind are:

1. Can the rest of the engine get into a bad state where it is trying to create a "one past the end" chunk range? If so, what is the root cause?
2. If not, is there a set of "valid" inputs to `nextChunkRange` that still causes the undesirable behavior?

**Comment by pierre-samsara** [comment]
Just stumbled upon this PR because I have encountered the exact same bug.
Based on the time frame I'm selecting I get this error when `SplitUntil` is activated.
Error:
`[2023-01-25 23:00:00 +0000 UTC..2023-01-25 23:00:00 +0000 UTC): InvalidParameterException: End time cannot be less than Start time (Service: AWSLogs; Status Code: 400; Error Code: InvalidParameterException;`

If I specify smaller chunks, the bug doesn't happen (I'm thinking it's because it doesn't need to do any split when the chunks are small enough).
It also seems like the bug is only present when the parallelism level is greater than 1.

**Comment by vcschapp** [comment]
Thank you @pierre-samsara!

I've been struggling to pinpoint it the actual logical cause. Would you be willing to share your query parameters and a log dump?

Query parameters would be just the contents of your `incite.QuerySpec{}`.

The log trace you can get by hooking up a logger to the `QueryManager` when you create it, just set the `Logger` field within `incite.Config{}`. The `Logger` interface is compatible with the standard `log` package so you should be able to just set it to `log.Default()`.

I'm convinced that with these data we can figure it out.

**Comment by pierre-samsara** [comment]
Here is the code I use to run the query:
```
	s, err := m.Query(incite.QuerySpec{
		Groups: []string{"gatewaymicrohubserver"},
		Text: `fields tomillis(@timestamp) AS timestamp, gatewayId, level
| filter level='error' AND msg='http error: unexpected EOF' OR msg='successfully received status in payload'`,
		Start: time.Date(2023, 01, 25, 19, 00, 00, 0, time.UTC),
		End:   time.Date(2023, 01, 25, 23, 00, 00, 0, time.UTC),

		Chunk:      10 * time.Minute,
		SplitUntil: 10 * time.Second,
		Limit:      incite.MaxLimit,
	})
```

Here is the the log dump with the bug happening:
[bug_run.log](https://github.com/gogama/incite/files/10521413/bug_run.log)

Hope this helps!

**Comment by vcschapp** [comment]
Thanks! Would you say that in your experience it is deterministic (or at least reliably happens with the same parameters) or is it closer to be a random occurrence?

**Comment by artificial-aidan** [comment]
I experienced both. Some parameter sets were deterministic, some were random. The random ones seemed to be when it was auto resizing the query

**Comment by vcschapp** [comment]
Interesting. In this case it looks like it's probably deterministic.

You're starting with `10`-minute chunks over a four-hour period from 19:00 to 23:00, so there should be 24 such chunks, with the last one being from 22:50 to 23:00, but for some reason Incite decided to create a 25th chunk (chunk ID `24`) :

> 2023/01/27 09:18:34 incite: QueryManager(0xc000152f00) permanent failure to start chunk 24 "fields tomillis(@timestamp) AS timestamp, gatewayId, level\n| filter level='error' AND msg='http error: unexpected EOF' OR msg='successfully received status in payload'" [2023-01-25 23:00:00 +0000 UTC..2023-01-25 23:00:00 +0000 UTC): fatal error from CloudWatch Logs: InvalidParameterException: End time cannot be less than Start time (Service: AWSLogs; Status Code: 400; Error Code: InvalidParameterException; Request ID: 3fd9b905-f518-4e21-8eed-2578a82f9321; Proxy: null)

It superficially seems to be the difference between `<=` and `<`  except if that were the case it would happen all the time.

**Comment by vcschapp** [comment]
OK, found the problem.

Yes, it is caused by chunk splitting. And I can see it why it would only happen with `Parallel > 1`.

The problem is [here](https://github.com/gogama/incite/blob/main/mgr.go#L411-L414). Basically splitting chunks causes `mgr.n` to increase, which depending on the vicissitudes of parallelism can result in `mgr.next < mgr.n` and the stream being pushed back into the priority queue in cases where the are no more chunks available to lazily create. So the next chunk from `stream.nextChunkRange()` ends up having an empty time range because of [this code](https://github.com/gogama/incite/blob/main/stream.go#L132-L134).

Probably the right thing to do is create a new field `mgr.last` to do the job `mgr.n` was originally meant to do, namely provide the constant number of original chunks requested before splitting. Chunk splitting caused `mgr.n` to stealthily take on a second duty and that should be reversed.

**Comment by vcschapp** [comment]
I pushed a bug fix to the branch `main`, so you can consume it now by consuming from the HEAD of `main` instead of the latest published release.

The fix should be included in the next release, v1.4.0, which is probably 3-4 weeks away. @artificial-aidan, @pierre-samsara, let me know if consuming from HEAD works for you for now. If not, we can cut a temporary release that just includes the fix.

**Comment by vcschapp** [comment]
Instead of waiting for v1.4.0, I created a mini-release [v1.3.4](https://github.com/gogama/incite/releases/tag/v1.3.4) which just adds this one bug fix to v1.3.3.

It should be available to consume now.

@artificial-aidan, @pierre-samsara, let me know if this new version seems to fix the problem for you.

**Comment by pierre-samsara** [comment]
@vcschapp It seems like the bug is fixed for me with the latest version 🙌
I've attached the logs for the query I ran to check the bug fix.
[bug_fix_run.log](https://github.com/gogama/incite/files/10538122/bug_fix_run.log)

Thanks for fixing this so quickly!

**Comment by vcschapp** [comment]
Awesome!!! I'm so glad we could turn it around so fast. Many thanks for sharing the logs and query params, they were invaluable to root causing.

**Pull request closed without merging**